## Data Analysis I

Lastly, we will answer the following questions using pandas methods: 

* What are the most popular colors by season?
* What is the most popular clothing item by season?
* What is the effect of promo-codes on the dollar-amount of purchases?
* When do users leave a review?
* Do states vary in the items they purchase in the Fall?

Utilize the documentation provided in each code-block. When you are done with this section of the project, validate that your output matches the screenshot provided in the `docs/part3.md` file and answer the questions located underneath `Data Analysis II` in your own words.

In [101]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from scipy import stats

In [102]:
# TODO: load `data/processed/shopping_cleaned.csv` as a pandas dataframe

df=pd.read_csv("../data/processed/shopping_cleaned.csv")

In [103]:
# TODO: Create a dataframe pivot table where "Season" is your column, "Color" is your index, and "Location" is your value (this value is arbitrary)
# use the len function as the "aggfunc" parameter.
# Save this pivot table to a new variable
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html

pivot_df = pd.pivot_table(df, values='Location', index='Color', columns='Season', aggfunc=len)

In [104]:
# TODO: Display this pivot table

pivot_df

Season                Fall  Spring  Summer  Winter
Color                                             
Aubergine               83       2      13     111
Baby blue               16      69      84      21
Black                   84       5       9     124
Brick red               88       7      12     110
Brown                   97       5       3      89
Burnt orange            93       4       5     112
Fuchsia                  3      45      78      18
Lavender                13      55     102       6
Lemon yellow             9      47      98      13
Mauve                   86       5       7      95
Muted mustard yellow    84       8      13      87
Pale peach               9      52      81      15
Periwinkle               9      54      82       8
Ruby Red                90       3       8     107
Terra cotta             91       4      15     101
Turquoise                7      49      74      16
White                   10      63      81      11

In [105]:
# TODO: Select the "Fall" column from this pivot table and display the sort order in descending order
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.sort_values.html

pivot_df['Fall'].sort_values(ascending=False)


Color
Brown                   97
Burnt orange            93
Terra cotta             91
Ruby Red                90
Brick red               88
Mauve                   86
Black                   84
Muted mustard yellow    84
Aubergine               83
Baby blue               16
Lavender                13
White                   10
Periwinkle               9
Pale peach               9
Lemon yellow             9
Turquoise                7
Fuchsia                  3
Name: Fall, dtype: int64

In [106]:
# TODO: Select the "Winter" column from this pivot table and display the sort order in descending order

pivot_df['Winter'].sort_values(ascending=False)

Color
Black                   124
Burnt orange            112
Aubergine               111
Brick red               110
Ruby Red                107
Terra cotta             101
Mauve                    95
Brown                    89
Muted mustard yellow     87
Baby blue                21
Fuchsia                  18
Turquoise                16
Pale peach               15
Lemon yellow             13
White                    11
Periwinkle                8
Lavender                  6
Name: Winter, dtype: int64

In [107]:
# TODO: Select the "Spring" column from this pivot table and display the sort order in descending order

pivot_df['Spring'].sort_values(ascending=False)

Color
Baby blue               69
White                   63
Lavender                55
Periwinkle              54
Pale peach              52
Turquoise               49
Lemon yellow            47
Fuchsia                 45
Muted mustard yellow     8
Brick red                7
Mauve                    5
Brown                    5
Black                    5
Burnt orange             4
Terra cotta              4
Ruby Red                 3
Aubergine                2
Name: Spring, dtype: int64

In [108]:
# TODO: Select the "Summer" column from this pivot table and display the sort order in descending order

pivot_df['Summer'].sort_values(ascending=False)

Color
Lavender                102
Lemon yellow             98
Baby blue                84
Periwinkle               82
Pale peach               81
White                    81
Fuchsia                  78
Turquoise                74
Terra cotta              15
Muted mustard yellow     13
Aubergine                13
Brick red                12
Black                     9
Ruby Red                  8
Mauve                     7
Burnt orange              5
Brown                     3
Name: Summer, dtype: int64

In [109]:
# TODO: Using the "idmax" method, get the index with the maximum value for each column
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html

pivot_df.idxmax()


Season
Fall          Brown
Spring    Baby blue
Summer     Lavender
Winter        Black
dtype: object

In [110]:
# TODO: Create a dataframe pivot table where "Season" is your column, "Item Purchased" is your index, and "Location" is your value (this value is arbitrary)
# use the len function as the "aggfunc" parameter.
# Save this pivot table to a new variable and display it

pivot_df_2 = pd.pivot_table(df, values='Location', index='Item Purchased', columns='Season', aggfunc=len)
pivot_df_2

Season          Fall  Spring  Summer  Winter
Item Purchased                              
Backpack        72.0     NaN     NaN   113.0
Belt            25.0     NaN     NaN     NaN
Boots           51.0     NaN     NaN    83.0
Dress           36.0    74.0   118.0     NaN
Gloves          68.0     NaN     NaN   112.0
Handbag         31.0    78.0   111.0     NaN
Hoodie          58.0     NaN     NaN   105.0
Jacket          47.0     NaN     NaN    87.0
Jeans           46.0     NaN     NaN     NaN
Leggings        58.0     NaN     NaN   118.0
Pajamas         65.0     NaN     NaN   111.0
Running Shoes   25.0    89.0   134.0     NaN
Scarf           56.0     NaN     NaN   101.0
Shorts          31.0    76.0   152.0     NaN
Socks           71.0     NaN     NaN   101.0
Sunglasses      43.0    80.0   135.0     NaN
Sweater         72.0     NaN     NaN   113.0
T-shirt         17.0    80.0   115.0     NaN

In [111]:
# TODO: Using the "idmax" method, get the index with the maximum value for each column on the pivot table containing clothing items & season
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.idxmax.html

pivot_df_2.idxmax()

Season
Fall           Backpack
Spring    Running Shoes
Summer           Shorts
Winter         Leggings
dtype: object

In [112]:
# TODO: Create two data-frames for promo-code users and non-promo-code users using boolean indexing
# Save these two dataframes into seperate variables 
# Documentation: https://pandas.pydata.org/docs/user_guide/indexing.html#boolean-indexing
# Hint: Users that have used a promo code have the value "Yes" in the column "Promo Code Used"

promo_users_df = df[df['Promo Code Used'] == 'Yes']

non_promo_users_df = df[df['Promo Code Used'] != 'Yes']

In [113]:
# TODO: Print the shape of the promo code users dataframe. This should be (1373, 13)

promo_users_df.shape

(1373, 13)

In [114]:
# TODO: Print the average purchase size of promo code users dataframe. This should be 50.06747998236351

promo_users_df["Purchase Amount (USD)"].mean()

50.06747998236351

In [115]:
# TODO: Print the shape of the non promo code users dataframe. This should be (1785, 13)

non_promo_users_df.shape

(1785, 13)

In [116]:
# TODO: Print the average purchase size of non promo code users dataframe. This should be 30.155750765104113

non_promo_users_df["Purchase Amount (USD)"].mean()

30.155750765104113

Formulate a null hypothesis regarding the dollar amount spent between promo-code users & non-promo-code users

There is no difference regarding the dollar amount spent between promo-code users and non-promo-code users. 

Formulate an alternative hypothesis regarding the dollar amount spent between promo-code users & non-promo-code users

There is a difference regarding the dollar amount spent between promo-code users and non-promo-code users. 

In [117]:
# TODO: Using the "ttest_ind" method from "scipy.stats", run a T-Test between the dollar amount spent of promo-code users and non promo-code users
# The t-statistic you should observe is 93.2464714350831
# Documentation: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html

stats.ttest_ind(promo_users_df["Purchase Amount (USD)"], non_promo_users_df["Purchase Amount (USD)"])

TtestResult(statistic=93.2464714350831, pvalue=0.0, df=3156.0)

In [118]:
# TODO: Group your dataframe according to the "Review Rating" column. This should produce two groups for the values "Missing" & "Present"
# Save this grouped dataframe into a new variable named "g_review"
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html

g_review = df.groupby(by= 'Review Rating')

In [119]:
# TODO: Print out the keys of this dataframe for validation. This should output "dict_keys(['Missing', 'Present'])"
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.keys.html

g_review.groups.keys()

dict_keys(['Missing', 'Present'])

In [120]:
# TODO: calculate the average of the "Purchase Amount (USD)", "Age", and "Previous Purchases" columns in our grouped dataframe
# output the values for display
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.mean.html
# Documentation: https://pandas.pydata.org/docs/getting_started/intro_tutorials/03_subset_data.html

g_review.mean("Purchase Amount (USD)", "Age", "Previous Purchases")

Age  Purchase Amount (USD)  Previous Purchases
Review Rating                                                      
Missing        34.080198              38.923623            5.344059
Present        34.371705              38.615947            6.778559

In [124]:
# TODO: Create a new column called "Prev Purchase Group" on your original dataframe which seperates your dataframe into 
# two groups of shoppers based on their "Previous Purchases" column using the "pd.cut" method. 
# Ensure that you are only creating 2 "bins" and label these respective bins as ["Low", "High"]
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.cut.html

df['Prev Purchase Group'] = pd.cut(df['Previous Purchases'], bins= 2, labels= ['Low','High'])
df['Prev Purchase Group']

0        Low
1        Low
2        Low
3        Low
4        Low
        ... 
3153    High
3154     Low
3155    High
3156     Low
3157     Low
Name: Prev Purchase Group, Length: 3158, dtype: category
Categories (2, object): ['Low' < 'High']

In [139]:
# TODO: Using the "value_counts" function, count how many Missing & Present values are in the "Low" group
# Display this value for analysis 
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html
# Hint: You will have to use Boolean Indexing

low_group = df[df['Prev Purchase Group'] == 'Low']
low_group['Review Rating'].value_counts() 

Review Rating
Missing    1686
Present     840
Name: count, dtype: int64

In [140]:
# TODO: Using the "value_counts" function, count how many Missing & Present values are in the "High" group
# Display this value for analysis 
# Documentation: https://pandas.pydata.org/docs/reference/api/pandas.Series.value_counts.html
# Hint: You will have to use Boolean Indexing

high_group = df[df['Prev Purchase Group'] == 'High']
high_group['Review Rating'].value_counts() 

Review Rating
Missing    334
Present    298
Name: count, dtype: int64

## Data Analysis II

In the next section, answer the primary analytical questions in your own words.

### Q1

What are the top three colors for Fall & Winter?

The top three colors for Fall were: Brown, Burnt Orange, and Terra Cotta

The top three colors for Winter were: Black, Burnt Orange, and Aubergine

It can be seen that the two seasons share Burnt Orange as the second most popular color for their respective seasons. However further analysis may need to be done to determine all three of the most popular colors for both seasons altogether. 

### Q2

What are the top three colors for Spring & Summer?

The top three colors for Spring were: Baby Blue, White, and Lavender

The top three colors for Summer were: Lavender, Lemon Yellow, and Baby Blue

It can be seen that the two seasons share Baby Blue and Lavender as their most popular colors. However further analysis may need to be done to determine all three of the most popular colors for both seasons altogether. 

### Q3

What is the most popular clothing item by season?

By season the most popular clothing items are as follows:

Fall           Backpack
Spring    Running Shoes
Summer           Shorts
Winter         Leggings

### Q4

Observing the p-value that we got when running our t-test between promo-code and non-promo-code users, what can we conclude regarding our null hypothesis?

Since the p-value was less that 0.05, we can reject null hypothesis and say that it is not true that there is no difference regarding the dollar amount spent between promo code and non-promo code users.

### Q5

Observe the `value_counts` output for the "Review Rating" column for both your "High" and "Low" groups. Proportionally speaking, which group is more likely to leave a review? Why might this be happening from the "human"-perspective? Rationalizations are ok at this point, even if they aren't backed up by data.

The high group was more likely to leave a review as they had 298 reviews out of the 632 total reviews within the table, giving about a 47% reviews present while the low group had about 33% reviews present. This can be because consumers who have made many purchases in the past with a company feel more compelled to leave a review due to incentives that can be apparent from them being a rewards member of the company for awhile. Thus, leaving a review as a long time customer may provide them discounts for more shopping.